### ISSUE TITLE
Wasm br_table spec running forever

### ISSUE THREAD

**Issue by bjfish** [bytecodealliance/cranelift#1243]
The `br_table.wast` spec appears to run forever when the `jump_tables_enabled` flag is set to false.  This may be a regression in recent versions like `0.49.0` as this was not observed in `0.44.0`.

- What are the steps to reproduce the issue? 
I used wasmtime at commit `a2479df32932b2481ae576a9b07a55081288b20c` to run the spectests.
   1. Add `flag_builder.set("jump_tables_enabled", "false").unwrap();` to `wast_testsuites.rs` in wasmtime.
  2. Run `cargo test br_table` in wasmtime.
  3. The test appears to run forever.

- What do you expect to happen? 
  The spectest should run to completion. 
- What does actually happen?
  The spectest appears to run forever.
- Which Cranelift version / commit hash / branch are you using?
`0.50.0`
- If relevant, can you include some extra information about your environment?
  (Rust version, operating system, architecture...)
rustc 1.39.0 (4560ea788 2019-11-04)
macOS Catalina

**Comment by data-pup** [comment]
Confirming that this happens in [lucet](https://github.com/bytecodealliance/lucet) as well.

### What are the steps to reproduce the issue?

Patch `compiler.rs` to disable jump tables, diff is shown below.

```
diff --git i/lucetc/src/compiler.rs w/lucetc/src/compiler.rs
index b5f80c4..b8d3ebd 100644
--- i/lucetc/src/compiler.rs
+++ w/lucetc/src/compiler.rs
@@ -246,6 +246,7 @@ impl<'a> Compiler<'a> {
         flags_builder.enable("enable_verifier").unwrap();
         flags_builder.enable("is_pic").unwrap();
         flags_builder.set("opt_level", opt_level.to_flag()).unwrap();
+        flags_builder.set("jump_tables_enabled", "false").unwrap();
         Ok(isa_builder.finish(settings::Flags::new(flags_builder)))
     }
 }
```

Run the `br_table` spectest using the following command:

```sh
cargo test -p lucet-spectest br_table
```

### Which Cranelift version / commit hash are you using?
`3217e4e3253bd15a93f1a191345c5e3a0dab31de` (tag: v0.46.1) 

I tried stepping back to find where this might have been introduced, but I was able to reproduce this issue as far back as `05d327316a7460d32832c1482c957e1d1b5c2d19` in `lucet`, which was then using Cranelift v0.40.0, (commit `5534fd2c2f411dcdb981f992239fc0e96790312b`) so this might actually be a longstanding issue.

Hope this helps!

**Comment by data-pup** [comment]
I poked around at this a little more today, and found some info that might help this issue get solved!

To start with, I narrowed down part of the file that was causing this test to
hang for so long when jump tables were disabled to
[this](https://github.com/WebAssembly/spec/blob/2e2e3a9e471d162ce317b6d1153e8ff8b9945218/test/core/br_table.wast#L107) block in the module.

If you shorten the label vector in that spec file to one row, this test actually completes fairly quickly when I run lucet's spectest suite. So, I performed an (extremely unscientific) benchmarking exercise on my laptop by running this command:

```
time cargo test -p lucet-spectest br_table
```

I edited the file to have an increasingly large number of rows between iterations. Each row contains 34 values, in case you were curious. That ended up giving me these results:

```
1 row:    0.14s user 0.02s system 94% cpu 0.170 total
2 rows:   0.20s user 0.03s system 100% cpu 0.234 total
4 rows:   0.30s user 0.03s system 97% cpu 0.340 total
8 rows:   0.65s user 0.03s system 98% cpu 0.686 total
16 rows:  1.72s user 0.02s system 100% cpu 1.740 total
32 rows:  5.50s user 0.04s system 99% cpu 5.562 total
64 rows:  19.44s user 0.03s system 99% cpu 19.466 total
128 rows: 72.52s user 0.03s system 99% cpu 1:12.56 total
256 rows: 286.00s user 0.04s system 99% cpu 4:46.05 total
```

This would be using the debug build I believe, but again, this was not a scientific exercise in benchmarking. The main point of interest here is that running this code will _eventually_ terminate, but that is a very long "eventually," given that the spec file linked above has 724 rows in the label vector given to `br_table`.

So, if the code will terminate, maybe this test is hanging in `wasmtime` and `lucet` when jump tables are disabled because codegen is getting bogged down? I poked around with `clif-util` to see (a) how long it takes to compile, and how large the resulting module is.

```
data-pup@cranelift $: time ./target/debug/clif-util wasm --set jump_tables_enabled=false --target=x86_64 -X small_table.wat
Function #0 code size: 509 bytes
Function #0 bytecode size: 55 bytes
Total module code size: 509 bytes
Total module bytecode size: 55 bytes

0.07s user 0.00s system 99% cpu 0.070 total

data-pup@cranelift $: time ./target/debug/clif-util wasm --set jump_tables_enabled=false --target=x86_64 -X big_table.wat
Function #0 code size: 165001 bytes
Function #0 bytecode size: 8726 bytes
Total module code size: 165001 bytes
Total module bytecode size: 8726 bytes

173.89s user 0.01s system 99% cpu 2:53.91 total

data-pup@cranelift $: time ./target/debug/clif-util wasm --target=x86_64 -X small_table.wat
Function #0 code size: 187 bytes
Function #0 bytecode size: 55 bytes
Total module code size: 187 bytes
Total module bytecode size: 55 bytes

0.01s user 0.00s system 96% cpu 0.013 total

data-pup@cranelift $: time ./target/debug/clif-util wasm --target=x86_64 -X big_table.wat
Function #0 code size: 34870 bytes
Function #0 bytecode size: 8726 bytes
Total module code size: 34870 bytes
Total module bytecode size: 8726 bytes

0.12s user 0.00s system 99% cpu 0.120 total
```

I also used `clif-util` to print the resulting Cranelift IR, if this is useful. Setting `jump_tables_enabled` didn't seem affect this output.

```
; Exported as "large"
function u0:0(i32, i64 vmctx) -> i32 system_v {
    jt0 = jump_table [ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2,
ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3, ebb2, ebb3]

                                ebb0(v0: i32, v1: i64):
@002a                               br_table v0, ebb2, jt0

                                ebb3:
@0052                               v3 = iconst.i32 0
@0054                               return v3

                                ebb2:
@0056                               v4 = iconst.i32 1
@0058                               return v4
}
```

**tldr;** Compilation seems to get _very_ slow when jump tables are disabled, and large vectors are given to a `br_table` instruction.

Note that this was tested on the following versions of Cranelift/Lucet:
  * Cranelift commit hash: `c54fc845`
  * Lucet commit hash: `4c75de4`

And for reference, the `br_table` spectest file can be found [here.](https://github.com/WebAssembly/spec/blob/2e2e3a9e471d162ce317b6d1153e8ff8b9945218/test/core/br_table.wast)
